lc-hax/Scripts/Patches/GodModePatch.cs

```diff
@@ -1,10 +1,125 @@
+    static bool PrefixJesterKill(int playerId, ref bool ___inKillAnimation) {
+        if (!Setting.EnableGodMode) return true;
+        if (Helper.IsLocalPlayerAboutToGetKilledByEnemy(playerId)) {
+            ___inKillAnimation = false;
+            return false;
+        }
+
+        return true;
+    }
+
+    [HarmonyPrefix]
+    [HarmonyPatch(typeof(MaskedPlayerEnemy), nameof(MaskedPlayerEnemy.KillPlayerAnimationServerRpc))]
+    [HarmonyPatch(typeof(MaskedPlayerEnemy), nameof(MaskedPlayerEnemy.KillPlayerAnimationServerRpc))]
+    static bool PrefixMaskedPlayerKill(int playerObjectId, ref bool ___startingKillAnimationLocalClient) {
+        if (!Setting.EnableGodMode) return true;
+        if (Helper.IsLocalPlayerAboutToGetKilledByEnemy(playerObjectId)) {
+            ___startingKillAnimationLocalClient = false;
+            return false;
+        }
+
+        return true;
+    }
+
+    [HarmonyPrefix]
+    [HarmonyPatch(typeof(MouthDogAI), nameof(MouthDogAI.KillPlayerServerRpc))]
+    [HarmonyPatch(typeof(MouthDogAI), nameof(MouthDogAI.KillPlayerClientRpc))]
+    static bool PrefixDogKill(MouthDogAI __instance, int playerId, ref bool ___inKillAnimation) {
+        if (!Setting.EnableGodMode) return true;
+        if (Helper.IsLocalPlayerAboutToGetKilledByEnemy(playerId)) {
+            ___inKillAnimation = false;
+            __instance.StopKillAnimationServerRpc();
+            return false;
+        }
+
+        return true;
+    }
+
+    // needed because the dogs do break the player even on contact, which makes it impossible to grab items,
+    // we patch it to make a controlled interaction for this, avoiding the death under god mode.
+    [HarmonyPrefix]
+    [HarmonyPatch(typeof(MouthDogAI), nameof(MouthDogAI.OnCollideWithPlayer))]
+    static bool ControlledDogCollision(MouthDogAI __instance, ref Collider other) {
+        if (!Setting.EnableGodMode) return true;
+        if (__instance.IsLocalPlayerAboutToGetKilledByEnemy(other))
+            if (Helper.LocalPlayer != null) {
+                // instead let's make it chase the player, but never kill it.
+                if (__instance.currentBehaviourStateIndex is 0 or 1) {
+                    _ = __instance.Reflect().InvokeInternalMethod("ChaseLocalPlayer");
+                }
+                else {
+                    if (__instance.currentBehaviourStateIndex != 2 ||
+                        __instance.Reflect().GetInternalField<bool>("inLunge"))
+                        return false;
+                    __instance.transform.LookAt(Helper.LocalPlayer.transform.position);
+                    __instance.transform.localEulerAngles = new Vector3(0.0f, Helper.LocalPlayer.transform.eulerAngles.y, 0.0f);
+                    _ = __instance.Reflect().SetInternalField("inLunge", true);
+                    _ = __instance.Reflect().InvokeInternalMethod("EnterLunge");
+                }
+
+                return false;
+            }
+
+        return true;
+    }
```

Comment:
I don't understand why you are not patching `KillPlayerServerRpc` instead. Did you test that it was this method that's causing the problem?

Reply:
yeah, each enemy fatal kill needs to be blocked to avoid getting the player bugged somewhere.

Reply:
It's bugged because of this line.

```cs
playerControllerB.inAnimationWithEnemy = this;
```

```cs
    public override void OnCollideWithPlayer(Collider other)
    {
        base.OnCollideWithPlayer(other);
        PlayerControllerB playerControllerB = MeetsStandardPlayerCollisionConditions(other, inKillAnimation);
        if (!(playerControllerB != null))
        {
            return;
        }

        Vector3 vector = Vector3.Normalize((base.transform.position + Vector3.up - playerControllerB.gameplayCamera.transform.position) * 100f);
        if (Physics.Linecast(base.transform.position + Vector3.up + vector * 0.5f, playerControllerB.gameplayCamera.transform.position, out var hitInfo, StartOfRound.Instance.collidersAndRoomMask, QueryTriggerInteraction.Ignore))
        {
            if (!(hitInfo.collider == debugCollider))
            {
                Debug.Log("Eyeless dog collide, linecast obstructed: " + hitInfo.collider.gameObject.name);
                debugCollider = hitInfo.collider;
            }
        }
        else if (currentBehaviourStateIndex == 3)
        {
            playerControllerB.inAnimationWithEnemy = this;
            KillPlayerServerRpc((int)playerControllerB.playerClientId);
        }
        else if (currentBehaviourStateIndex == 0 || currentBehaviourStateIndex == 1)
        {
            ChaseLocalPlayer();
        }
        else if (currentBehaviourStateIndex == 2 && !inLunge)
        {
            base.transform.LookAt(other.transform.position);
            base.transform.localEulerAngles = new Vector3(0f, base.transform.eulerAngles.y, 0f);
            inLunge = true;
            EnterLunge();
        }
    }
```

All you need to do is add a prefix that sets `currentBehaviourStateIndex` to not 3.

Reply:
Yeah, but id rather want the enemy do the lunge animation, which is why behaviour 3 is set.

Reply:
Lunge animation is 2 not 3.

Reply:
enum MouthDog {
    ROAMING = 0,
    SUSPICIOUS = 1,
    CHASE = 2,
    LUNGE = 3
}
no, is 3.

Reply:
The DLL literally says it's 2. 3 is only set AFTER entering a lunge.

Reply:
setting the RPC to 3 it goes in lunge animation.

Reply:
No, it doesn't. It just enrages the dog.

Reply:
```cs
            case 3:
                if (base.IsOwner)
                {
                    agent.speed -= Time.deltaTime * 5f;
                    if (!endingLunge && agent.speed < 1.5f && !inKillAnimation)
                    {
                        endingLunge = true;
                        lungeCooldown = 0.25f;
                        EndLungeServerRpc();
                    }
                }

                break;
```

It represents the end of a lunge.

Reply:
```cs
            case 2:
                if (!hasEnteredChaseModeFully)
                {
                    if (!startedChaseModeCoroutine)
                    {
                        startedChaseModeCoroutine = true;
                        StartCoroutine(enterChaseMode());
                    }

                    break;
                }

                ((RigConstraint<DampedTransformJob, DampedTransformData, DampedTransformJobBinder<DampedTransformData>>)(object)neckDampedTransform).weight = Mathf.Lerp(((RigConstraint<DampedTransformJob, DampedTransformData, DampedTransformJobBinder<DampedTransformData>>)(object)neckDampedTransform).weight, 0.2f, 8f * Time.deltaTime);
                if (!base.IsOwner)
                {
                    break;
                }

                if (base.IsOwner && roamPlanet.inProgress)
                {
                    StopSearch(roamPlanet);
                }

                if (!inLunge)
                {
                    lungeCooldown -= Time.deltaTime;
                    if (Vector3.Distance(base.transform.position, noisePositionGuess) < 4f && lungeCooldown <= 0f)
                    {
                        inLunge = true;
                        EnterLunge();
                        break;
                    }
                }

                agent.speed = Mathf.Clamp(agent.speed + Time.deltaTime, 13f, 18f);
                if (stunNormalizedTimer > 0f)
                {
                    agent.speed = 0f;
                }

                AITimer -= Time.deltaTime;
                if (AITimer <= 0f)
                {
                    AITimer = 3f;
                    suspicionLevel--;
                    if (Vector3.Distance(base.transform.position, agent.destination) < 3f)
                    {
                        SearchForPreviouslyHeardSound();
                    }

                    if (suspicionLevel <= 8)
                    {
                        SwitchToBehaviourState(1);
                    }
                }

                break;
```

2 is the lunge.

Reply:
2 makes the dog run, 3 makes it do the ambush animation, and if that hits a player is calling the kill RPC, you can literally test it out with possessionmod!

Reply:
I did test with PossessionMod. 3 does makes it do a roar. Not a lunge.

Reply:
Feel free to make any changes you think are the best ones.

Reply:
It's alright. I can wait.